job-server/src/spark.jobserver/JobCache.scala

```diff
@@ -22,28 +26,51 @@ class JobCacheImpl(maxEntries: Int,
                    loader: ContextURLClassLoader) extends JobCache {
   import scala.concurrent.duration._
 
-  private val cache = new LRUCache[(String, DateTime, String), JobJarInfo](maxEntries)
+  private val cache = new LRUCache[(String, DateTime, String), BinaryJobInfo](maxEntries)
   private val logger = LoggerFactory.getLogger(getClass)
   implicit val daoAskTimeout: Timeout = Timeout(3 seconds)
 
   /**
    * Retrieves the given SparkJob class from the cache if it's there, otherwise use the DAO to retrieve it.
-   * @param appName the appName under which the jar was uploaded
-   * @param uploadTime the upload time for the version of the jar wanted
+   * @param appName the appName under which the binary was uploaded
+   * @param uploadTime the upload time for the version of the binary wanted
    * @param classPath the fully qualified name of the class/object to load
    */
   def getSparkJob(appName: String, uploadTime: DateTime, classPath: String): JobJarInfo = {
-    cache.get((appName, uploadTime, classPath), {
-      import akka.pattern.ask
-      import scala.concurrent.Await
-
-      val jarPathReq = (dao ? JobDAOActor.GetJarPath(appName, uploadTime)).mapTo[JobDAOActor.JarPath]
-      val jarPath = Await.result(jarPathReq, daoAskTimeout.duration).jarPath
-      val jarFilePath = new java.io.File(jarPath).getAbsolutePath()
-      sparkContext.addJar(jarFilePath) // Adds jar for remote executors
-      loader.addURL(new URL("file:" + jarFilePath)) // Now jar added for local loader
-      val constructor = JarUtils.loadClassOrObject[spark.jobserver.api.SparkJobBase](classPath, loader)
-      JobJarInfo(constructor, classPath, jarFilePath)
-    })
+    cache.get((appName, uploadTime, classPath)) match {
```

Comment:
The cache is already futures based!   Read up on the spray cache.   That's
why I recommend sticking with the existing method of passing in a function
as the second argument of get(...)....  which creates the future to
populate the cache if the key cannot be found, and ensures other threads
gets that same future to avoid the stampeding herds problem.

On Thu, Aug 11, 2016 at 4:14 PM, Stephen Carman <notifications@github.com>
wrote:

> In job-server/src/spark.jobserver/JobCache.scala
> <https://github.com/spark-jobserver/spark-jobserver/pull/553#discussion_r74520877>
> :
>
> >     * @param classPath the fully qualified name of the class/object to load
> >     */
> >    def getSparkJob(appName: String, uploadTime: DateTime, classPath: String): JobJarInfo = {
> > -    cache.get((appName, uploadTime, classPath), {
> > -      import akka.pattern.ask
> > -      import scala.concurrent.Await
> > -
> > -      val jarPathReq = (dao ? JobDAOActor.GetJarPath(appName, uploadTime)).mapTo[JobDAOActor.JarPath]
> > -      val jarPath = Await.result(jarPathReq, daoAskTimeout.duration).jarPath
> > -      val jarFilePath = new java.io.File(jarPath).getAbsolutePath()
> > -      sparkContext.addJar(jarFilePath) // Adds jar for remote executors
> > -      loader.addURL(new URL("file:" + jarFilePath)) // Now jar added for local loader
> > -      val constructor = JarUtils.loadClassOrObject[spark.jobserver.api.SparkJobBase](classPath, loader)
> > -      JobJarInfo(constructor, classPath, jarFilePath)
> > -    })
> > +    cache.get((appName, uploadTime, classPath)) match {
>
> Ideally, we should rewrite the cache entirely to be future based and just
> pass futures around on this kind of stuff.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/spark-jobserver/spark-jobserver/pull/553/files/b4ea3b476442b188582c22f86fe5e18e504de3b4#r74520877>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABA323Y64ogKrm13dtU8bb-TTN8b4I7gks5qe6zogaJpZM4JZq4L>
> .
>



-- 
If you are free, you need to free somebody else.
If you have some power, then your job is to empower somebody else.
   --- Toni Morrison